---
title: "eQTM: Correlating methylation and expression"
author: "Andrew Y.F. Li Yim"
date: "June/July 2017"
output:
  pdf_document: 
    highlight: tango
    number_sections: yes
    toc: yes
---

Step 1.2 Expression Quantitative Trait Methylation Analysis
===============================
In this workbook we aim to perform an expression quantitative trait methylation analysis, which we seek to perform by correlating the DMRs with the expression of the genes associated in step 1.1. The approach implemented here is to calculate the average Beta per DMR per sample obtained from the WGBS data and to correlate this value with the smoothed count data obtained from the RNAseq data for the samples that were present in both experiments. To prevent outliers from influencing our data too much, we will calculate the ranked Pearson correlation coefficient (aka Spearman correlation).

# Importing the annotations
Change the enhancer_folder and promoter_folder if necessary to reflect the most up to date annotations. 
```{r Folders}
work_folder <- "/home/al580162/projects/BTCURE/"
enhancer_folder <- paste0(work_folder, "1_Reannotate_DMRs/Results/DMenhancers/Separate/Annotations/170710") 
promoter_folder <- paste0(work_folder, "1_Reannotate_DMRs/Results/DMpromoters/Separate/Annotations/170704")

#EPI server
dmr_folder <- paste0(work_folder, "0_Original_project_local/WGBS")
exprs_folder <- paste0(work_folder, "0_Original_project_local/RNAseq")

#CB servers
dmr_folder <- paste0(work_folder, "0_Original_project/data/analysis/WGBS/07differentialMethylation")
exprs_folder <- paste0(work_folder, "0_Original_project/data/analysis/RNAseq/07countReads")

```

## Importing the differentially methylated enhancers
```{r Importing the DMenhancers}
require(GenomicRanges)

HvL_dmrs_enhancers <- read.csv(file.path(enhancer_folder, "HvL_dmenhancers.csv"), stringsAsFactors = F)[, -1] 
HvL_dmrs_enhancers$dmrChr <- gsub("chr", "", as.character(HvL_dmrs_enhancers$dmrChr))
HvL_dmrs_enhancers$contrast <- "HvL"

LvN_dmrs_enhancers <- read.csv(file.path(enhancer_folder, "LvN_dmenhancers.csv"), stringsAsFactors = F)[, -1]
LvN_dmrs_enhancers$dmrChr <- gsub("chr", "", as.character(LvN_dmrs_enhancers$dmrChr))
LvN_dmrs_enhancers$contrast <- "LvN"

HvN_dmrs_enhancers <- read.csv(file.path(enhancer_folder, "HvN_dmenhancers.csv"), stringsAsFactors = F)[, -1]
HvN_dmrs_enhancers$dmrChr <- gsub("chr", "", as.character(HvN_dmrs_enhancers$dmrChr))
HvN_dmrs_enhancers$contrast <- "HvN"

dm_enhancers <- rbind(HvL_dmrs_enhancers, LvN_dmrs_enhancers, HvN_dmrs_enhancers) 
dm_enhancers$celltype <- gsub(", *", ";", dm_enhancers$celltype)
dm_enhancers <- aggregate(contrast~, dm_enhancers, FUN = toString)

dm_enhancers_gr <- keepStandardChromosomes(makeGRangesFromDataFrame(dm_enhancers, keep.extra.columns = T, seqnames.field = "dmrChr", start.field = "dmrStart", end.field = "dmrEnd"))
dm_enhancers_unique_gr <- unique(dm_enhancers_gr)
```

## Importing the differentially methylated enhancers
```{r Importing the DMpromoters}
HvL_dmrs_promoters <- read.csv(file.path(promoter_folder, "HvL_dmpromoters.csv"), stringsAsFactors = F)[, -1]
HvL_dmrs_promoters$dmrChr <- gsub("chr", "", as.character(HvL_dmrs_promoters$dmrChr))
HvL_dmrs_promoters$contrast <- "HvL"

HvN_dmrs_promoters <- read.csv(file.path(promoter_folder, "HvN_dmpromoters.csv"), stringsAsFactors = F)[, -1]
HvN_dmrs_promoters$dmrChr<- gsub("chr", "", as.character(HvN_dmrs_promoters$dmrChr))
HvN_dmrs_promoters$contrast <- "HvN"

dm_promoters <- rbind(HvL_dmrs_promoters, HvN_dmrs_promoters)
dm_promoters$celltype <- gsub(", *", ";", dm_promoters$celltype)
dm_promoters_gr <- keepStandardChromosomes(makeGRangesFromDataFrame(dm_promoters, keep.extra.columns = T, seqnames.field = "dmrChr", start.field = "dmrStart", end.field = "dmrEnd"))
dm_promoters_unique_gr <- unique(dm_promoters_gr)
```

```{r Making one big GRanges object of all unique DMRs}
dm_enhancers_gr_unified <- dm_enhancers_gr[, c("baitENS", "baitSYMBOL", "nCpGs", "direction")]
colnames(mcols(dm_enhancers_gr_unified)) <- c("ENS", "SYMBOL", "nCpGs", "direction")
dm_enhancers_gr_unified$REG <- "Enhancer"

dm_promoters_gr_unified <- dm_promoters_gr[, c("geneENS", "geneSYMBOL", "nCpGs", "direction")]
colnames(mcols(dm_promoters_gr_unified)) <- c("ENS", "SYMBOL", "nCpGs", "direction")
dm_promoters_gr_unified$REG <- "Promoter"

dm_gr <- c(dm_enhancers_gr_unified, dm_promoters_gr_unified) 
dm_gr <- makeGRangesFromDataFrame(unique(data.frame(dm_gr)), keep.extra.columns = T)
dm_gr_unique <- unique(dm_gr)
```

Which DMR gene pairs are present several times in our dataset
```{r Promiscuous DMRs}
unique_dmr_genes <- table(paste0(as.character(seqnames(dm_gr)), ":", start(dm_gr), "-", end(dm_gr)))
promiscuous_dmrs <- names(which(unique_dmr_genes > 1))
promiscuous_dmrs_df <- data.frame(chr = gsub("(.+):.+-.+", "\\1", promiscuous_dmrs),
				  start = gsub(".+:(.+)-.+", "\\1", promiscuous_dmrs), 
 				  end = gsub(".+:.+-(.+)", "\\1", promiscuous_dmrs))
promiscuous_dmrs_gr <- makeGRangesFromDataFrame(promiscuous_dmrs_df)
promiscuous_dmrs_gr_anno <- sort((subsetByOverlaps(dm_gr, promiscuous_dmrs_gr)))
length(promiscuous_dmrs_gr)

```
In summary, we find 82 DMRs that are present in regulatory regions that have a priori been found to regulate several genes.

# Preparing the methylation data
## Importing the BSSeq data
The data imported and processed previously were based on the DMR coordinates, for the purpose of correlation we will need to extract the Beta value for each CpG comprising the DMR. As the actual matrix of all Beta values will be very large, best is to subset the data else the averaging might choke up.
```{r Importing the BSSeq data}
require(bsseq)

fit2 <- readRDS(file.path(dmr_folder, "fit2.rds"))
fit2 <- keepStandardChromosomes(fit2, pruning.mode = "coarse")
colnames(fit2) <- pData(fit2)$Sample

fit2_dmrs <- subsetByOverlaps(fit2, dm_gr_unique)
```

##Averaging the DMRs
```{r Averaging the Betas per DMR per sample}
require(parallel)
no_cores <- 6 

dmr_beta_mean <- function(dmr_gene, methylation, ...){
     if(is.null(dmr_gene)) stop("No GRanges object with DMRs provided")
     if(is.null(methylation)) stop("No GRanges object with methylation data provided")

     dmr_chr <- gsub(pattern = "chr", replacement = "", x = as.character(seqnames(dmr_gene)), ignore.case = T)

     meth_chr <- as.character(seqnames(methylation))
     meth_pos <- start(methylation)

     if(class(methylation) == "BSseq"){
          betas <- bsseq::getMeth(methylation[which(meth_chr == dmr_chr & meth_pos >= start(dmr_gene) & meth_pos <= end(dmr_gene)), ], type = "raw")
     } else if(class(methylation) == "GenomicRanges"){
          betas <- methylation[which(meth_chr == dmr_chr & meth_pos >= start(dmr_gene) & meth_pos <= end(dmr_gene)), ]
          betas <- as.data.frame(betas)
     } else stop("The methylation matrix must be a GRanges(-derived) object")
     betas <- colMeans(betas, na.rm = T)
     return(betas)
}
```

```{r Averaging the DMRs}
gc()
cl <- makeCluster(no_cores)
clusterExport(cl, c("dmr_beta_mean", "fit2_dmrs"))
clusterEvalQ(cl, c(library(GenomicRanges),
		   library(bsseq)))
dm_mean <- parLapply(cl = cl, X = dm_gr_unique, fun = function(dmr){dmr_beta_mean(dmr_gene = dmr, methylation = fit2_dmrs)})
stopCluster(cl)
gc()

dm_mean_df <- data.frame(matrix(unlist(dm_mean), ncol = ncol(fit2_dmrs), byrow = T))
rownames(dm_mean_df) <- paste0(seqnames(dm_gr_unique), ":", start(dm_gr_unique), "-", end(dm_gr_unique))
colnames(dm_mean_df) <- pData(fit2_dmrs)$Sample

na_dm_indices <- which(is.na(dm_mean_df), arr.ind = T)[,1]
if(length(na_dm_indices) != 0){
	dm_mean_df <- dm_mean_df[-na_dm_indices, ]
}
dm_mean_df$chr <- gsub("(.+):.+-.+", "\\1", rownames(dm_mean_df))
dm_mean_df$start <- gsub(".+:(.+)-.+", "\\1", rownames(dm_mean_df))
dm_mean_df$end <- gsub(".+:.+-(.+)", "\\1", rownames(dm_mean_df))
dm_mean_gr <- makeGRangesFromDataFrame(dm_mean_df, keep.extra.columns = T, seqnames.field = "chr", start.field = "start", end.field = "end")

dm_gr_culled <- subsetByOverlaps(dm_gr, dm_mean_gr)
#As the data is composed of DMRs from different contrasts, we will need to find the unique sequences lest we accidentally run a correlative analysis twice (waste of CPU time)
dm_gr_culled_unique <- unique(split(dm_gr_culled, dm_gr_culled$SYMBOL))
#Reduce dm_gr_culled_unique to merge all the overlapping DMRs
#dm_gr_culled_merged <- reduce(dm_gr_culled_unique)


```

# Preparing the expression data
## Importing the RNASeq data
```{r Importing the RNASeq data}
#expr_data <- read.csv(file.path(exprs_folder, "counts.txt"), sep = "\t")
#colnames(expr_data) <- gsub("\\.", "-", colnames(expr_data))
#expr_data <- read.csv(file.path(exprs_folder, "normCounts4.txt"), sep = "\t")
#colnames(expr_data) <- gsub("\\.", "-", colnames(expr_data))
expr_data <- read.csv(file.path(work_folder, "1_Reannotate_DMRs/Data/RNAseq/rld.csv"), row.names = 1)
colnames(expr_data) <- gsub("(RA)\\.([0-9]+).+", "\\1-\\2", colnames(expr_data))

expr_data <- expr_data[which(rowSums(expr_data) != 0), ]

HvL_degs <- read.csv(file.path(exprs_folder, "../08differentialExpression/HL.deg.csv"), stringsAsFactors = F)
HvL_degs <- with(HvL_degs, HvL_degs[order(pvalue), ])
HvL_degs_sig <- HvL_degs[HvL_degs$padj < 0.05, ]
HvL_degs_sig <- HvL_degs_sig[-unique(which(is.na(HvL_degs_sig), arr.ind = T)[, 1]), ]

HvN_degs <- read.csv(file.path(exprs_folder, "../08differentialExpression/HN.deg.csv"), stringsAsFactors = F)
HvN_degs <- with(HvN_degs, HvN_degs[order(pvalue), ])
HvN_degs_sig <- HvN_degs[HvN_degs$padj < 0.05, ]
HvN_degs_sig <- HvN_degs_sig[-unique(which(is.na(HvN_degs_sig), arr.ind = T)[, 1]), ]

LvN_degs <- read.csv(file.path(exprs_folder, "../08differentialExpression/LN.deg.csv"), stringsAsFactors = F)
LvN_degs <- with(LvN_degs, LvN_degs[order(pvalue), ])
LvN_degs_sig <- LvN_degs[LvN_degs$padj < 0.05, ]
LvN_degs_sig <- LvN_degs_sig[-unique(which(is.na(LvN_degs_sig), arr.ind = T)[, 1]), ]

degs <- unique(c(HvL_degs_sig$ENSEMBL, HvN_degs_sig$ENSEMBL, LvN_degs_sig$ENSEMBL))

```

# Perform the eQTM analysis
## eQTM analysis
```{r eQTM analysis}
require(boot)
require(parallel)

#The reason the correlator core is separate from the bootstrapper and the permuter is because the bootstrapper function can be parallelized using parLapply
correlator_core <- function(dmr_data, expr_data, dmr_gene_anno, id_col_name, meth_data, iterations, cor_type = c("pearson", "kendall", "spearman"), ncores = 1, seed = NULL){
	#dmr_data: A GRanges object whose coordinates correspond to the coordinates of the DMR and whose metadata represents the aggregated methylation value per sample. Can be generated using the dmr_beta_mean.
	#expr_data: A dataframe containing transcription with a transcript/gene identifier as row names and sample names as column names. Note that the sample names must overlap with the sample names of dmr_data.
	#dmr_gene_anno: A GRanges object whose coordinates correspond to the coordinates of the DMR and whose metadata contains a column containing the transcript/gene identifier associated to the DMR.
	#id_col_name: The name of the metadata column of dmr_gene_anno that contains the transcript/gene identifier.
	#meth_data: A GRanges(-derived) object (such as BSseq) necessary for the permutation analysis 
	#iterations: An integer representing the number of bootstraps and permutations to perform.
	#cor_type: A string representing which type of correlation to calculate ("pearson", "spearman", "kendall").
	#ncores: An integer representing the number of cores to use for parallelization. Defaults to the singlecore implementation.
	#seed: A value to seed the randomization process. Used for reproducibility purposes.

	if(is.null(dmr_data)) stop("dmr_data cannot be found")
	if(class(dmr_data) != "GRanges") stop("dmr_data must be a GRanges object")
	if(is.null(expr_data)) stop("expr_data cannot be found")
	if(!any(colnames(mcols(dmr_data)) %in% colnames(expr_data))) stop("There exists no overlap between the samples present in dmr_data and expr_data as evident from the column names")
	if(is.null(id_col_name)) stop("id_col_name cannot be found")
	if(is.null(meth_data)) stop("meth_data cannot be found")
	if(!class(meth_data) %in% c("BSseq", "GRanges")) stop("meth_data must be a GRanges or BSseq object")
	if(is.null(iterations)) stop("iterations cannot be found")

	iterations <- round(iterations)
	cor_type <- match.arg(cor_type)

	if(!is.numeric(ncores)) stop("ncores must be a numeric")
	if(ncores > 1){
		cat(paste0("Parallelization will be implemented using ", ncores, " cores.\n"))
		#Parallelization under UNIX environments works better if forked (better memory management). However, forking is not an option under windows
		if(Sys.info()['sysname'] != "Windows"){
			cl_type <- "FORK"
		} else{
			cl_type <- "PSOCK"
		}
	} else cat(paste0("Computation will be performed on a single core.\n"))

	#Basic QC expression
	expr_data <- expr_data[rowSums(expr_data) != 0, ]

	#Find overlapping samples
	dmr_samples <- as.character(colnames(mcols(dmr_data)))
	expr_samples <- as.character(colnames(expr_data))
	overlapping_samples <- intersect(dmr_samples, expr_samples)

	#Find DMR-annotations for which aggregated methylation values exist
	overlapping_dmr_gene_anno <- subsetByOverlaps(dmr_gene_anno, dmr_data)
	#Find overlapping transcription features
	overlapping_dmr_gene_anno <- overlapping_dmr_gene_anno[mcols(overlapping_dmr_gene_anno)[, id_col_name] %in% rownames(expr_data), ]

	dmr_data_culled <- dmr_data[, overlapping_samples]
	expr_data_culled <- expr_data[, overlapping_samples]
	
	cat("Pulling up Baron Munchausen by his bootstraps\n")
	if(ncores == 1){
		if(!is.null(seed)) set.seed(seed)
		correlations <- sapply(X = overlapping_dmr_gene_anno, simplify = T, USE.NAMES = F, FUN = function(dmr_gene_entry){
			cor_df <- data.frame(meth = as.vector(unlist(mcols(subsetByOverlaps(dmr_data_culled, dmr_gene_entry)))),
					     expr = as.vector(unlist(expr_data_culled[unlist(mcols(dmr_gene_entry[, id_col_name])), ])))
			ci_bootstrapper(cor_df = cor_df, iterations = iterations, cor_type = cor_type)
		})
	} else{
		cl <- makeCluster(spec = ncores, type = cl_type)
		clusterExport(cl, c("dmr_data_culled", "id_col_name", "iterations", "cor_type", "ci_bootstrapper"), envir = environment())
		clusterEvalQ(cl, library(boot))
		if(!is.null(seed)) clusterSetRNGStream(cl = cl, iseed = seed)
			      
		correlations <- parSapply(cl = cl, X = overlapping_dmr_gene_anno, simplify = T, USE.NAMES = F, FUN = function(dmr_gene_entry){
			cor_df <- data.frame(meth = as.vector(unlist(mcols(subsetByOverlaps(dmr_data_culled, dmr_gene_entry)))),
					     expr = as.vector(unlist(expr_data_culled[unlist(mcols(dmr_gene_entry[, id_col_name])), ])))
			ci_bootstrapper(cor_df = cor_df, iterations = iterations, cor_type = cor_type)
		})
		stopCluster(cl)
		gc()
	}
		
	correlations <- t(correlations)

	correlations_df <- cbind(data.frame(overlapping_dmr_gene_anno), correlations)
	correlations_gr <- makeGRangesFromDataFrame(correlations_df, keep.extra.columns = T)
	correlations_gr$nCpGs <- countOverlaps(correlations_gr, meth_data)

	return(correlations_gr)
}

ci_bootstrapper <- function(cor_df, iterations, cor_type){
	#Bootstraps for the CI
	eqtm_correlator <- function(df, indices){
		return(cor(df[indices, "meth"], df[indices, "expr"], method = cor_type)) 
	}
	boot_results <- boot(cor_df, eqtm_correlator, R = iterations, stype = "i")
	CI95 <- boot.ci(boot_results, type = "bca")

	return(c(rho = boot_results$t0, rhoCI95_lower = CI95$bca[4], rhoCI95_upper = CI95$bca[5]))
}

pval_permuter <- function(correlations_gr, meth_data, expr_data, iterations = 1000, alternative = c("two.sided", "greater", "less"), id_col_name, cor_type, ncores = 1, seed = NULL){
	if(is.null(correlations_gr)) stop("correlations_gr cannot be found")
	if(class(correlations_gr) != "GRanges") stop("correlations_gr must be a GRanges object")
	if(is.null(meth_data)) stop("meth_data cannot be found")
	if(!class(meth_data) %in% c("BSseq", "GRanges")) stop("meth_data must be a GRanges or BSseq object")
	if(is.null(colnames(meth_data))) stop("meth_data must have sample names as column names")
	if(is.null(expr_data)) stop("expr_data cannot be found")
	if(is.null(id_col_name) | !id_col_name %in% colnames(mcols(correlations_gr))) stop("Cannot find the expression ID column in correlations_gr")
	if(is.null(cor_type)) stop("cor_type cannot be found")
	if(!is.numeric(ncores)) stop("cores must be a numeric")
	if(ncores > 1){
		#Parallelization under UNIX environments works better if forked (better memory management). However, forking is not an option under windows
		if(Sys.info()['sysname'] != "Windows"){
			cat("Parallelization using FORK\n")
			cl_type <- "FORK"
		} else{
			cat("Parallelization using PSOCK\n")
			cl_type <- "PSOCK"
		}
	}
	
	overlapping_samples <- intersect(colnames(meth_data), colnames(expr_data))
	if(length(overlapping_samples) == 0 ) stop("No overlapping samples found for meth_data and expr_data")

	## Generate null distribution
	permutation_df <- unique(data.frame(chr = as.character(seqnames(correlations_gr)),
				     CpGs = correlations_gr$nCpGs))
	permutation_df <- permutation_df[order(permutation_df$chr, permutation_df$CpGs), ]
	chromosome_list <- with(permutation_df, split(permutation_df, chr))

	#Split per chromosome for computational purposes
	permuted_cor <- lapply(names(chromosome_list), function(chromosome){
		cat(paste0("Starting on chromosome ", chromosome, ".\n"))   
		chr_data <- meth_data[which(seqnames(meth_data) == chromosome), ]
		
		dmr_length <- chromosome_list[[chromosome]]
		
		#Find a random set of $nCpG indices $iterations times
		if(ncores == 1){
			if(!is.null(seed)) set.seed(seed)
			permuted_cor <- lapply(dmr_length$CpGs, function(CpG){
				eff_nrow <- nrow(chr_data)-as.numeric(CpG)
				indices <- sample(x = 1:eff_nrow, size = iterations, replace = T)

				nulldist <- t(sapply(X = indices, simplify = T, FUN = function(index){
				       colMeans(getMeth(chr_data[index:(index+CpG-1), overlapping_samples], type = "raw"), na.rm = T)
				}))
				return(nulldist)
	    		})
		} else{
			cl <- makeCluster(spec = ncores, type = cl_type)
			if(!is.null(seed)) clusterSetRNGStream(cl = cl, iseed = seed)
			clusterExport(cl, c("chr_data", "iterations", "overlapping_samples"), envir = environment())
			clusterEvalQ(cl, library(bsseq))

			permuted_cor <- parLapply(cl = cl, X = dmr_length$CpGs, fun = function(CpG){
				eff_nrow <- nrow(chr_data)-as.numeric(CpG)
				indices <- sample(x = 1:eff_nrow, size = iterations, replace = T)

				nulldist <- t(sapply(X = indices, simplify = T, FUN = function(index){
				       colMeans(getMeth(chr_data[index:(index+CpG-1), overlapping_samples], type = "raw"), na.rm = T)
				}))
				return(nulldist)
	    		})
			stopCluster(cl)
			gc()

		}
		names(permuted_cor) <- paste0(dmr_length$chr, "_", dmr_length$CpGs)
		return(permuted_cor)
	})
	permuted_cor <- unlist(permuted_cor, recursive = F)
	permuted_cor_chr <- gsub("(.+)_.+", "\\1", names(permuted_cor))
	permuted_cor_nCpGs <- gsub(".+_(.+)", "\\1", names(permuted_cor))

	## Calculate probabilities
	cat("Calculating the probabilities under the null (aka \"pvalues\")\n")
	if(ncores == 1){

		pvals <- lapply(correlations_gr, function(cor_entry){

			perm_dmrs <- unlist(permuted_cor[[which(permuted_cor_chr == seqnames(cor_entry) & permuted_cor_nCpGs == cor_entry$nCpGs)]][, overlapping_samples])
			transcription <- unlist(expr_data[unlist(mcols(cor_entry)[id_col_name]), overlapping_samples])

			null_rhodist <- apply(X = perm_dmrs, MARGIN = 1, FUN = function(perm_dmr){
				cor(perm_dmr, transcription, method = cor_type)
			})

			if(alternative == "two.tailed"){
				pval <- mean(abs(cor_entry$rho) > abs(null_rhodist), na.rm = T)
			} else if(alternative == "greater"){
				pval <- mean(cor_entry$rho > null_rhodist, na.rm = T)
			} else if(alternative == "less"){
				pval <- mean(cor_entry$rho < null_rhodist, na.rm = T)
			}
		       return(pval)	
		})
	} else{
		cl <- makeCluster(spec = ncores, type = cl_type)
		if(!is.null(seed)) clusterSetRNGStream(cl = cl, iseed = seed)
		clusterExport(cl, c("permuted_cor", "permuted_cor_chr", "permuted_cor_nCpGs", "overlapping_samples", "expr_data", "id_col_name"), envir = environment())

		pvals <- parLapply(cl = cl, X = correlations_gr, fun = function(cor_entry){

			perm_dmrs <- unlist(permuted_cor[[which(permuted_cor_chr == seqnames(cor_entry) & permuted_cor_nCpGs == cor_entry$nCpGs)]][, overlapping_samples])
			transcription <- unlist(expr_data[unlist(mcols(cor_entry)[id_col_name]), overlapping_samples])

			null_rhodist <- apply(X = perm_dmrs, MARGIN = 1, FUN = function(perm_dmr){
				cor(perm_dmr, transcription, method = cor_type)
			})

			if(alternative == "two.tailed"){
				pval <- mean(abs(cor_entry$rho) < abs(null_rhodist), na.rm = T)
			} else if(alternative == "greater"){
				pval <- mean(cor_entry$rho < null_rhodist, na.rm = T)
			} else if(alternative == "less"){
				pval <- mean(cor_entry$rho > null_rhodist, na.rm = T)
			}
			return(pval)	
		})
	}
	stopCluster(cl)
	gc()
	correlations_gr$pvals <- unlist(pvals)

	correlations_gr <- correlations_gr[with(correlations_gr, order(pvals, rev(abs(rho)), decreasing = F)), ]
	
	return(correlations_gr)
}

dm_gr_correlations <- correlator_core(dmr_data = dm_mean_gr, expr_data = expr_data, meth_data = fit2, dmr_gene_anno = dm_gr_culled_unique, id_col_name = "ENS", iterations = 1000, cor_type = "spearman", ncores = 2, seed = 124718)

dm_gr_correlations <- pval_permuter(correlations_gr = dm_gr_correlations, meth_data = fit2, expr_data = expr_data, iterations = 1000, id_col_name = "ENS", cor_type = "spearman", ncores = 2, alternative = "two.tailed", seed = 124718)
dm_gr_correlations$padj <- p.adjust(dm_gr_correlations$pvals)

write.csv(data.frame(dm_gr_correlations), "eqtms.csv")

deqtms <- dm_gr_correlations[dm_gr_correlations$ENS %in% degs,]
write.csv(data.frame(deqtms), "deqtms.csv")

```

Do we observe any correlation between the correlation coefficient and the p-value
```{r Distribution correlation coefficient versus p value}
require(ggExtra)
require(ggplot2)
cor_sig_plotter <- function(eqtm_gr){
	cell_df <- with(eqtm_gr, data.frame(rho = rho,
					    log10p = -log10(pvals),
					    feature = REG))
	plot_obj <- ggplot(cell_df, aes(x = rho, y = log10p)) +
			geom_point(aes(col = feature)) + 
			ylab("-log10(p)") +
			theme_bw() +
			xlim(-1, 1) 
	return(plot_obj)	
}

png(file = "corvsig.png", width = 800, height = 800, res = 150)
print(cor_sig_plotter(dm_gr_correlations))
dev.off()

enhancer_eqtms <- dm_gr_correlations[dm_gr_correlations$REG == "Enhancer", ]
png(file = "enhancer_cordist.png", width = 800, height = 800, res = 150)
enh_plot <- ggplot(as.data.frame(enhancer_eqtms), aes(x = rho)) + 
		   xlim(-1, 1) +
		   geom_histogram(aes(y = ..density..)) + 
		   geom_density() +
		   theme_bw()
print(enh_plot)	   
dev.off()

promoter_eqtms <- dm_gr_correlations[dm_gr_correlations$REG == "Promoter", ]
png(file = "promoter_cordist.png", width = 800, height = 800, res = 150)
prom_plot <- ggplot(as.data.frame(promoter_eqtms), aes(x = rho)) +
		    xlim(-1, 1) +
		    geom_histogram(aes(y = ..density..)) +
		    geom_density() +
		    theme_bw()
print(prom_plot)
dev.off()

```



We are curious whether certain DMRs/genes are enriched for TFBS/pathways.
```{r subsetting for enrichment analyses}
eqtm_enhancers <- dm_gr_correlations[which(dm_gr_correlations$REG == "Enhancer"), ]
eqtm_promoters <- dm_gr_correlations[which(dm_gr_correlations$REG == "Promoter"), ]


hypo_neg_dmen <- neg_dmen[which(neg_dmen$dmrdirection == "hypo"), ]
hyper_neg_dmen <- neg_dmen[which(neg_dmen$dmrdirection == "hyper"), ]

pos_dmen <- dmen_corr_df[which(dmen_corr_df$pval < 0.05 & dmen_corr_df$rho > 0), ]
hypo_pos_dmen <- pos_dmen[which(pos_dmen$dmrdirection == "hypo"), ]
hyper_pos_dmen <- pos_dmen[which(pos_dmen$dmrdirection == "hyper"), ]

neg_dmprom <- dmprom_corr_df[which(dmprom_corr_df$pval < 0.05 & dmprom_corr_df$rho < 0), ]
hypo_neg_dmprom <- neg_dmprom[which(neg_dmprom$dmrdirection == "hypo"), ]
hyper_neg_dmprom <- neg_dmprom[which(neg_dmprom$dmrdirection == "hyper"), ]

pos_dmprom <- dmprom_corr_df[which(dmprom_corr_df$pval < 0.05 & dmprom_corr_df$rho > 0), ]
hypo_pos_dmprom <- pos_dmprom[which(pos_dmprom$dmrdirection == "hypo"), ]
hyper_pos_dmprom <- pos_dmprom[which(pos_dmprom$dmrdirection == "hyper"), ]

write.csv(hypo_neg_dmen, "hypo_neg_dmen.csv")
write.csv(hyper_neg_dmen, "hyper_neg_dmen.csv")
write.csv(hypo_pos_dmen, "hypo_pos_dmen.csv")
write.csv(hyper_pos_dmen, "hyper_pos_dmen.csv")

write.csv(hypo_neg_dmprom, "hypo_neg_dmprom.csv")
write.csv(hyper_neg_dmprom, "hyper_neg_dmprom.csv")
write.csv(hypo_pos_dmprom, "hypo_pos_dmprom.csv")
write.csv(hyper_pos_dmprom, "hyper_pos_dmprom.csv")

```



## Visualization of the eQTM
```{r eQTM visualization}
eqtm_plot <- function(dmr_data, expr_data, chr, start, end, ensembl_id, degrees, plot_title, legend_title = NULL){
	#dmr_data: A dataframe containing the methylation data with individual methylation features as rows and samples as columns.
	#expr_data: A dataframe containing the expression data with individual expression features as rows and samples as columns.
	#degrees: A vector containing the degrees used to color the points with. Names must be the same as the ones used for the methylation and expression data.

	overlapping_samples <- intersect(colnames(expr_data), colnames(mcols(dmr_data)) )

	degrees <- degrees[overlapping_samples]
	dmr_data <- unlist(as.data.frame(mcols(dmr_data[which(seqnames(dmr_data) == chr & start(dmr_data) == start & end(dmr_data) == end), overlapping_samples])))
	expr_data <- unlist(expr_data[ensembl_id, overlapping_samples])

	plot_df <- data.frame(methylation = dmr_data, transcription = expr_data, groups = degrees) 

	require(ggplot2)
	p <- ggplot(plot_df, aes(x = methylation, y = transcription)) + 
		geom_point(aes(fill = groups), color = "black", shape = 21, size = 5) + 
		theme_bw() +
		xlab("Methylation") + 
		ylab("Expression") + 
		xlim(0,1) +
		ggtitle(plot_title) + 
		theme(plot.title = element_text(face = "bold"),
			axis.title = element_text(size = 14, face = "bold"),
			axis.text = element_text(size = 12), 
			legend.title = element_text(size = 14, face = "bold"),
			legend.text = element_text(size = 12),
			legend.position = "bottom")
	if(is.numeric(plot_df$groups) & !is.null(legend_title)){
		p <- p + scale_fill_continuous(name = legend_title)
	} else{
		p <- p + scale_fill_discrete(name = legend_title)
	}
	return(p)
}

named_swelling <- pData(fit2)$Swelling
names(named_swelling) <- pData(fit2)$Sample
named_sjc <- pData(fit2)$SJC
names(named_sjc) <- pData(fit2)$Sample

png("tophit.png", height = 800, width = 800, res = 150)
eqtm_plot(dmr_data = dm_mean_gr, expr_data = expr_data, chr = as.character(seqnames(dm_gr_correlations_pval[5,])), start = start(dm_gr_correlations_pval[5,]), end = end(dm_gr_correlations_pval[5,]), degrees = named_swelling, ensembl_id = as.character(dm_gr_correlations_pval[5, ]$ENS), plot_title = "PDE4A")
dev.off()

png("EVI2B.png", height = 800, width = 800, res = 150)
eqtm_plot(dmr_data = dm_mean_gr, expr_data = expr_data, chr = as.character(seqnames(deqtms))[1], start = start(deqtms)[1], end = end(deqtms)[1], degrees = named_swelling, ensembl_id = deqtms$ENS[1], plot_title = deqtms$SYMBOL[1])
dev.off()

png("SP140.png", height = 800, width = 800, res = 150)
eqtm_plot(dmr_data = dm_mean_gr, expr_data = expr_data, chr = as.character(seqnames(deqtms))[40], start = start(deqtms)[40], end = end(deqtms)[40], degrees = named_swelling, ensembl_id = deqtms$ENS[40], plot_title = deqtms$SYMBOL[40])
dev.off()


eqtm_top_plot <- function(dmr_data, expr_data, eqtm_top, degrees, number = 1, legend_title){
	if(length(number) == 1){
		name_plot <- paste0(eqtm_top[number, "dmrcoords"], "_", eqtm_top[number, "ensembl"], " (", eqtm_top[number, "symbol"], ")")
		png(file = paste0(eqtm_top[number, "ensembl"], ".png"), width = 800, height = 800, res = 150)
		print(eqtm_plot(dmr_data = dmr_data, expr_data = expr_data, coordinates = as.character(eqtm_top[number, "dmrcoords"]), degrees = degrees, ensembl_id = as.character(eqtm_top[number, "ensembl"]), plot_title = name_plot, legend_title = legend_title))
		dev.off()
	} else{
		lapply(number, function(iterator){ 
			name_plot <- paste0(eqtm_top[iterator, "dmrcoords"], "_", eqtm_top[iterator, "ensembl"], " (", eqtm_top[iterator, "symbol"], ")")
			png(file = paste0(iterator, "_", eqtm_top[iterator, "ensembl"], ".png"), width = 800, height = 800, res = 150)
			print(eqtm_plot(dmr_data = dmr_data, expr_data = expr_data, coordinates = as.character(eqtm_top[iterator, "dmrcoords"]), degrees = degrees, ensembl_id = as.character(eqtm_top[iterator, "ensembl"]), plot_title = name_plot, legend_title = legend_title))
			dev.off()
		})
	}
}

# Top 10 eQTMs
eqtm_top_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, eqtm_top = dmen_corr, degrees = named_sjc, number = 1:10, legend_title = "SJC")
eqtm_top_plot(dmr_data = dmprom_betas_mean_df, expr_data = expr_data, eqtm_top = dmprom_corr, degrees = named_sjc, number = 1:10, legend_title = "SJC")

enhancer_celltypes <- unique(unlist(strsplit(as.character(dmen_corr$celltype), ";")))
rho_sig_plot(corr_df = dmen_corr, celltypes = enhancer_celltypes)
promoter_celltypes <- unique(gsub("_(C|V)B", "", unlist(strsplit(as.character(dmprom_corr$celltype), ";"))))
rho_sig_plot(corr_df = dmprom_corr, celltypes = promoter_celltypes)

plot_df <- dmen_corr %>% separate_rows(celltype, sep = "(,|;)")
png(file = "enhancer_rho_dist.png", width = 800, height = 800, res = 150)
ggplot(plot_df, aes(x = rho)) +
	geom_density(aes(fill = factor(celltype), alpha = 0.05)) +
	xlim(-1,1) +
	theme(legend.position = "bottom") +
	ggtitle("Enhancer rho distribution")
dev.off()

promoter_celltypes <- unique(unlist(strsplit(as.character(dmprom_corr$celltype), ";")))
rho_sig_plot(corr_df = dmprom_corr, celltypes = promoter_celltypes)

dmprom_corr2 <- dmprom_corr_df
dmen_corr2 <- dmen_corr_df
colnames(dmprom_corr2)[4:6] <- colnames(dmen_corr2)[4:6] <- c("regChr", "regStart", "regEnd") 
dmreg_corr <- rbind(dmprom_corr2, dmen_corr2)
dmreg_corr_gr <- makeGRangesFromDataFrame(dmreg_corr, keep.extra.columns = T, seqnames.field = "dmrChr", start.field = "dmrStart", end.field = "dmrEnd")
write.csv(dmreg_corr, "dmreg_corr.csv")
```

One of the interesting DMregs was found at 14:23366500-23366804. It is associated to SLC7A7, which was reported in association to RA previously.
```{r SLC7A7}
png("SLC7A7.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "14:23366500-23366804", "ENSG00000155465", factors = named_sjc, plot_name = "SLC7A7", legend_name = "SJC"))
dev.off()
```
# Specific questions to be answered
Now that we have the eQTM analysis results, we have a few questions of interest regarding the promoters and enhancers:
1. Which genes are regulated by several regulatory regions?
2. Which regulatory regions contain DMRs and affect multiple genes?

## Question 1: Which genes are regulated by several regulatory regions?
To answer this question, the fast way is to simply tabulate the number of gene symbol occurrences in the list
```{r Gene symbol summary}
sort(table(dm_gr_correlations$SYMBOL))
table(table(dm_gr_correlations$SYMBOL))
```
At first glance it appears as though the differentially methylated regulatory regions appear to have many genes that are regulated by several regulatory elements, which is what we expect. However, the picture is a bit exaggerated as we see that some DMRs are "non-unique". This was not picked up because the DMRs map to the same region but are not exactly the same. This also means that the associated statistics are slightly different. 
Our solution is to remove the duplicated entries per gene symbol/ensembl ID. While removing multiple entries for the same target gene symbol is not an ideal situation (i.e. we accidentally remove multiple non-overlapping DMRs within the same region), it gives us at the very least an idea of which genes are affected by DMRs in different regulatory regions. The entry that is removed right now is the entry that comes first. Later updates will have to change this into something more agreeable (i.e. longest/shortest DMR).
```{r Deduping overlapping DMRs regions}
prom_symbols <- names(which(table(dm_gr_correlations$SYMBOL) > 1))
prom_genes <- dm_gr_correlations[dm_gr_correlations$SYMBOL %in% prom_symbols, ]
prom_genes_list <- split(prom_genes, prom_genes$SYMBOL)
number_unique_promdmrs <- lapply(prom_genes_list, function(prom_gene){
				  reduced_gene <- reduce(prom_gene)
				  length(reduced_gene)
			})
true_prom_genes <- names(which(number_unique_promdmrs > 1))
true_prom_genes <- prom_genes_list[true_prom_genes]
true_prom_genes_df <- as.data.frame(unlist(true_prom_genes))
write.csv(true_prom_genes_df, "promgenes.csv")


```
Two DMRs were found in one enhancer of KAT6B, one of which was strongly correlated
```{r KAT6B}
png("KAT6B_dmr1_enhancer.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "10:75929820-75930111", "ENSG00000156650", degrees = named_sjc, plot_title = "DMR 1 in enhancer KAT6B", legend_title = "SJC"))
dev.off()

png("KAT6B_dmr2_enhancer.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "10:75928975-75929471", "ENSG00000156650", degrees = named_sjc, plot_title = "DMR 2 in enhancer KAT6B", legend_title = "SJC"))
dev.off()
```

RIPK2 contained 3 DMRs that were associated to two enhancer regions.
```{r RIPK2}
png("RIPK2_dmr1_enhancer1.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "8:92062566-92062609", "ENSG00000104312", degrees = named_sjc, plot_title = "DMR 1 in enhancer 1 RIPK2", legend_title = "SJC"))
dev.off()

png("RIPK2_dmr2_enhancer2.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "8:92078337-92078855", "ENSG00000104312", degrees = named_sjc, plot_title = "DMR 2 in enhancer 2 RIPK2", legend_title = "SJC"))
dev.off()

png("RIPK2_dmr3_enhancer2.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "8:92079556-92079883", "ENSG00000104312", degrees = named_sjc, plot_title = "DMR 2 in enhancer 2 RIPK2", legend_title = "SJC"))
dev.off()

```


Analysis showed a very interesting correlation for PAPD5, which was found to be regulated through its promoter and an enhancer. 
```{r PAPD5}
#Enhancer
png("PAPD5i_dmr1_enhancer.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "16:50203607-50205157", "ENSG00000121274", degrees = named_sjc, plot_title = "DMR 1 in enhancer PAPD5", legend_title = "SJC"))
dev.off()
#Promoter
png("PAPD5_dmr2_promoter.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmprom_betas_mean_df, expr_data = expr_data, coordinates = "16:50154665-50154991", "ENSG00000121274", degrees = named_sjc, plot_title = "DMR 2 in promoter PAPD5", legend_title = "SJC"))
dev.off()
```

Three hits were reported for 2 DMRs in 2 enhancers for SP140. It should be noted however that while the DMRs do not overlap, the enhancer regions are adjacent to one another (as was alluded to previously). 
```{r SP140}
png("SP140_dmr1_enhancer1.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "2:231541887-231541958", "ENSG00000079263", degrees = named_sjc, plot_title = "DMR 1 in enhancer 1 SP140", legend_title = "SJC"))
dev.off()
png("SP140_dmr2_enhancer1.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "2:231545170-231546134", "ENSG00000079263", degrees = named_sjc, plot_title = "DMR 2 in enhancer 2 SP140", legend_title = "SJC"))
dev.off()
```


Despite our best efforts we managed to find a DMR that resided within an enhancer for two genes, thereby answering question 2 partially. Surprisingly both genes appear to behave the same.
```{r DMR 1:181131123-181132025}
png("MR1.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "1:181131123-181132025", "ENSG00000153029", factors = named_sjc, plot_name = "MR1", legend_name = "SJC"))
dev.off()
png("IER5.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "1:181131123-181132025", "ENSG00000162783", factors = named_sjc, plot_name = "IERS", legend_name = "SJC"))
dev.off()

#Need to correlate the actual expression of MR1 and IER5
IER5_MR1 <- t(expr_data[c("ENSG00000153029", "ENSG00000162783"), ] )
summary(lm(IER5_MR1[,1] ~ IER5_MR1[,2]))
```

## Question 2: Which regulatory regions contain DMRs and affect multiple genes?
To answer the question which differentially methylated regulatory regions affect multiple genes, we can utilize the same strategy as the one implemented in question 1, namely sorting and splitting for the regulatory coordinates. 
```{r Regulatory region summary}
table(paste0(seqnames(dm_gr_correlations), start(dm_gr_correlations), end(dm_gr_correlations)))
table(table(paste0(seqnames(dm_gr_correlations), start(dm_gr_correlations), end(dm_gr_correlations))))

```

As some DMRs overlap with one another due to them being called through different contrast, the first step is to perform a reduce() and then group all the DMRs that overlap the same area.
```{r Finding the promiscuous DMregs}
eqtm_coords <- paste0(seqnames(dm_gr_correlations), start(dm_gr_correlations), end(dm_gr_correlations))
true_prom_dmrs <- split(dm_gr_correlations, eqtm_coords)
num_prom_dmrs <- lapply(true_prom_dmrs, function(entry){
	length(entry)
})
prom_dmrs <- true_prom_dmrs[which(num_prom_dmrs > 1)]

write.csv(data.frame(prom_dmrs), "promdmrs.csv") 
```

```{r Plotting the correlations for the promiscuous DMRs}
cl <- makeCluster(no_cores)
promdmrs_folder <- "promdmrs"
clusterExport(cl, varlist = c("genes_promdmrs", "dmen_betas_mean_df", "expr_data", "eqtm_plot", "promdmrs_folder", "named_sjc"))

dir.create(promdmrs_folder)
parLapply(cl = cl, names(genes_promdmrs), function(promdmr_group){
	promdmr_subfolder <- paste0(promdmrs_folder, "/", gsub(":", "_", promdmr_group))       
	dir.create(promdmr_subfolder)
        require(foreach)

	coords_genes <- unique(data.frame(genes_promdmrs[[promdmr_group]][,c("seqnames", "start", "end", "ensembl", "symbol")]))
	apply(coords_genes, 1, function(dmr){
		coords <- paste0(dmr[1], ":", dmr[2], "-", dmr[3])
		ens <- dmr[4]
		png(file.path(promdmr_subfolder, paste0("chr", dmr[1], "_", dmr[2], "-", dmr[3], "_", dmr[5], ".png")), width = 800, height = 800, res = 150)
		print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = coords, ensembl_id = ens, degrees = named_sjc, plot_title = paste0(coords, " ", dmr[5]), legend_title = "SJC"))
		dev.off()
	})
})
stopCluster(cl)

cor(t(expr_data[c("ENSG00000146094","ENSG00000131188","ENSG00000146083","ENSG00000169220","ENSG00000184840","ENSG00000183258","ENSG00000165671","ENSG00000027847"),]))

```


