---
title: "eQTM: Correlating methylation and expression"
author: "Andrew Y.F. Li Yim"
date: "June/July 2017"
output:
  pdf_document: 
    highlight: tango
    number_sections: yes
    toc: yes
---

Step 1.2 Expression Quantitative Trait Methylation Analysis
===============================
In this workbook we aim to perform an expression quantitative trait methylation analysis, which we seek to perform by correlating the DMRs with the expression of the genes associated in step 1.1. The approach implemented here is to calculate the average Beta per DMR per sample obtained from the WGBS data and to correlate this value with the raw count data obtained from the RNAseq data for the samples that were present in both experiments. To prevent outliers from influencing our data too much, we will calculate the ranked Pearson correlation coefficient (aka Spearman correlation).

# Importing the annotations
Change the enhancer_folder and promoter_folder if necessary to reflect the most up to date annotations. 
```{r Folders}
work_folder <- "/home/al580162/projects/BTCURE/"
enhancer_folder <- paste0(work_folder, "1_Reannotate_DMRs/Results/DMenhancers/Separate/Annotations/170710") 
promoter_folder <- paste0(work_folder, "1_Reannotate_DMRs/Results/DMpromoters/Separate/Annotations/170704")

dmr_folder <- paste0(work_folder, "0_Original_project/data/analysis/WGBS/07differentialMethylation")
exprs_folder <- paste0(work_folder, "0_Original_project/data/analysis/RNAseq/07countReads")

```

## Importing the differentially methylated enhancers
```{r Importing the DMenhancers}
require(GenomicRanges)

HvL_dmrs_enhancers <- read.csv(file.path(enhancer_folder, "HvL_dmenhancers.csv"), stringsAsFactors = F)[, -1] 
HvL_dmrs_enhancers$dmrChr <- gsub("chr", "", as.character(HvL_dmrs_enhancers$dmrChr))
HvL_dmrs_enhancers$contrast <- "HvL"
HvL_dmrs_enhancers_gr <- makeGRangesFromDataFrame(HvL_dmrs_enhancers, keep.extra.columns = T, seqnames.field = "dmrChr", start.field = "dmrStart", end.field = "dmrEnd")

LvN_dmrs_enhancers <- read.csv(file.path(enhancer_folder, "LvN_dmenhancers.csv"), stringsAsFactors = F)[, -1]
LvN_dmrs_enhancers$dmrChr <- gsub("chr", "", as.character(LvN_dmrs_enhancers$dmrChr))
LvN_dmrs_enhancers$contrast <- "LvN"
LvN_dmrs_enhancers_gr <- makeGRangesFromDataFrame(LvN_dmrs_enhancers, keep.extra.columns = T, seqnames.field = "dmrChr", start.field = "dmrStart", end.field = "dmrEnd")

HvN_dmrs_enhancers <- read.csv(file.path(enhancer_folder, "HvN_dmenhancers.csv"), stringsAsFactors = F)[, -1]
HvN_dmrs_enhancers$dmrChr <- gsub("chr", "", as.character(HvN_dmrs_enhancers$dmrChr))
HvN_dmrs_enhancers$contrast <- "HvN"

dm_enhancers <- rbind(HvL_dmrs_enhancers, LvN_dmrs_enhancers, HvN_dmrs_enhancers) 
dm_enhancers$celltype <- gsub(", *", ";", dm_enhancers$celltype)
dm_enhancers_gr <- keepStandardChromosomes(makeGRangesFromDataFrame(dm_enhancers, keep.extra.columns = T, seqnames.field = "dmrChr", start.field = "dmrStart", end.field = "dmrEnd"))
```

## Importing the differentially methylated enhancers
```{r Importing the DMpromoters}
HvL_dmrs_promoters <- read.csv(file.path(promoter_folder, "HvL_dmpromoters.csv"), stringsAsFactors = F)[, -1]
HvL_dmrs_promoters$dmrChr <- gsub("chr", "", as.character(HvL_dmrs_promoters$dmrChr))
HvL_dmrs_promoters$contrast <- "HvL"

HvN_dmrs_promoters <- read.csv(file.path(promoter_folder, "HvN_dmpromoters.csv"), stringsAsFactors = F)[, -1]
HvN_dmrs_promoters$dmrChr<- gsub("chr", "", as.character(HvN_dmrs_promoters$dmrChr))
HvN_dmrs_promoters$contrast <- "HvN"

dm_promoters <- rbind(HvL_dmrs_promoters, HvN_dmrs_promoters)
dm_promoters$celltype <- gsub(", *", ";", dm_promoters$celltype)
dm_promoters_gr <- keepStandardChromosomes(makeGRangesFromDataFrame(dm_promoters, keep.extra.columns = T, seqnames.field = "dmrChr", start.field = "dmrStart", end.field = "dmrEnd"))
```

# Preparing the methylation data
## Importing the BSSeq data
The data imported and processed previously were based on the DMR coordinates, for the purpose of correlation we will need to extract the Beta value for each CpG comprising the DMR. As the actual matrix of all Beta values will be very large, best is to subset the data else the averaging might choke up.
```{r Importing the BSSeq data}
require(bsseq)

fit2 <- readRDS(file.path(dmr_folder, "fit2.rds"))
fit2 <- keepStandardChromosomes(fit2, pruning.mode = "coarse")

dmen_fit2 <- subsetByOverlaps(fit2, dm_enhancers_gr)
dmen_betas <- getMeth(dmen_fit2, type = "raw")
rownames(dmen_betas) <- paste0(seqnames(dmen_fit2), "_", start(dmen_fit2))
colnames(dmen_betas) <- pData(dmen_fit2)$Sample

dmprom_fit2 <- subsetByOverlaps(fit2, dm_promoters_gr)
dmprom_betas <- getMeth(dmprom_fit2, type = "raw")
rownames(dmprom_betas) <- paste0(seqnames(dmprom_fit2), "_", start(dmprom_fit2))
colnames(dmprom_betas) <- pData(dmprom_fit2)$Sample
```

##Averaging the DMRs
```{r Averaging the Betas per DMR per sample}
require(parallel)
no_cores <- detectCores()-1

dmr_beta_mean <- function(dmr_gr, cpg_beta){
	cpg_chr <- gsub("([0-9XY]+)_[0-9]+", "\\1", rownames(cpg_beta))
	cpg_start <- gsub("[0-9XY]+_([0-9]+)", "\\1", rownames(cpg_beta))
	dmr_chr <- gsub("chr", "", as.character(seqnames(dmr_gr)))
	betas <- cpg_beta[which(cpg_chr == dmr_chr & cpg_start >= start(dmr_gr) & cpg_start <= end(dmr_gr)),]
	mean_beta <- colMeans(betas, na.rm = T)
	return(mean_beta)
}
```

###Enhancers
```{r Averaging the DMenhancers}
cl <- makeCluster(no_cores)
clusterExport(cl, c("dmr_beta_mean", "dmen_betas"))
dmen_betas_mean <- parLapply(cl = cl, X = dm_enhancers_gr, fun = function(dmr){dmr_beta_mean(dmr_gr = dmr, cpg_beta = dmen_betas)})
stopCluster(cl)

names(dmen_betas_mean) <- paste0(seqnames(dm_enhancers_gr), ":", start(dm_enhancers_gr), "-", end(dm_enhancers_gr))
dmen_betas_mean_df <- do.call(rbind, dmen_betas_mean)

na_dmen_indices <- which(is.na(dmen_betas_mean_df), arr.ind = T)[,1]
if(length(na_dmen_indices) != 0){
	dmen_betas_mean_df <- dmen_betas_mean_df[-na_dmen_indices, ]
	dm_enhancers_gr <- dm_enhancers_gr[-na_dmen_indices, ]
}
dmen_beta_gene <- data.frame(dmrcoords = as.character(rownames(dmen_betas_mean_df)), 
			     enhcoords = as.character(paste0(dm_enhancers_gr$oeChr, ":", dm_enhancers_gr$oeStart, "-", dm_enhancers_gr$oeEnd)),
			     ensembl = as.character(dm_enhancers_gr$baitENS), 
			     symbol = as.character(dm_enhancers_gr$baitSYMBOL), 
			     contrast = as.character(dm_enhancers_gr$contrast), 
			     celltype = as.character(dm_enhancers_gr$celltype),
			     stringsAsFactors = F)
dmen_beta_gene <- unique(dmen_beta_gene)
dmen_beta_gene <- aggregate(contrast~., dmen_beta_gene, FUN = toString)
dmen_beta_gene$contrast <- gsub(", *", ";", dmen_beta_gene$contrast)
```

###Promoters
```{r Averaging the DMenhancers}
cl <- makeCluster(no_cores)
clusterExport(cl, c("dmr_beta_mean", "dmprom_betas"))
dmprom_betas_mean <- parLapply(cl = cl, X = dm_promoters_gr, fun = function(dmr){dmr_beta_mean(dmr_gr = dmr, cpg_beta = dmprom_betas)})
stopCluster(cl)

names(dmprom_betas_mean) <- paste0(seqnames(dm_promoters_gr), ":", start(dm_promoters_gr), "-", end(dm_promoters_gr))
dmprom_betas_mean_df <- do.call(rbind, dmprom_betas_mean)

na_prom_indices <- which(is.na(dmprom_betas_mean_df), arr.ind = T)[,1]
if(length(na_prom_indices) != 0){
	dmprom_betas_mean_df <- dmprom_betas_mean_df[-na_prom_indices, ]
	dm_promoters_gr <- dm_promoters_gr[-na_prom_indices, ]
}
dmprom_beta_gene <- data.frame(dmrcoords = as.character(rownames(dmprom_betas_mean_df)), 
			       promcoords = as.character(paste0(dm_promoters_gr$promChr, ":", dm_promoters_gr$promStart, "-", dm_promoters_gr$promEnd)),
			       ensembl = as.character(dm_promoters_gr$geneENS), 
			       symbol = as.character(dm_promoters_gr$geneSYMBOL), 
			       contrast = as.character(dm_promoters_gr$contrast), 
			       celltype = as.character(dm_promoters_gr$celltype),
			       stringsAsFactors = F)
dmprom_beta_gene <- unique(dmprom_beta_gene)
dmprom_beta_gene <- aggregate(contrast~., dmprom_beta_gene, FUN = toString)
dmprom_beta_gene$contrast <- gsub(", *", ";", dmprom_beta_gene$contrast)
dmprom_beta_gene$ensembl <- as.character(dmprom_beta_gene$ensembl)
# Remove the overlapping DMRs
dmprom_beta_gene_split <- split(dmprom_beta_gene, dmprom_beta_gene$ensembl)
```

# Preparing the expression data
## Importing the RNASeq data
```{r Importing the BSSeq data}
expr_data <- read.csv(file.path(exprs_folder, "counts.txt"), sep = "\t")
colnames(expr_data) <- gsub("\\.", "-", colnames(expr_data))
```

# Perform the eQTM analysis
## eQTM analysis
```{r eQTM analysis}
eqtm_func <- function(dmr_data, expr_data, dmr_gene_annotation, bootstraps = 1000, env = environment(), seed = NULL){
	require(boot)

	expr_data <- expr_data[which(rowSums(expr_data) != 0),]

	overlapping_dmrs <- intersect(dmr_gene_annotation[,"dmrcoords"], rownames(dmr_data))
	overlapping_samples <- intersect(colnames(expr_data), colnames(dmr_data)) 
	overlapping_genes <- intersect(rownames(expr_data), as.character(dmr_gene_annotation$ensembl)) 

	dmr_gene_annotation_culled <- unique(dmr_gene_annotation[which(dmr_gene_annotation$dmrcoords %in% overlapping_dmrs &
								       dmr_gene_annotation$ensembl %in% overlapping_genes), ])
	
	correlator <- function(dmreg){
		cor_df <- data.frame(dmr = unlist(dmr_data[unlist(dmreg["dmrcoords"]), overlapping_samples]),
	  			     expr = unlist(expr_data[unlist(dmreg["ensembl"]), overlapping_samples]))

		#Bootstrap for the CI
		spcor <- function(df, indices){
			return(cor(df[indices, "dmr"], df[indices, "expr"], method = "spearman"))
		}
		bootresults <- boot(cor_df, spcor, R = bootstraps, stype = "i")
		if(bootresults$t0 < 0){
			pval <- mean(bootresults$t >= 0)
		}else{
			pval <- mean(bootresults$t <= 0)
		}

	        CI95 <- boot.ci(bootresults, type = "bca")	

		return(c(rho = bootresults$t0, rhoCI95_lower = CI95$bca[4], rhoCI95_upper = CI95$bca[5], pval = pval))
	}

# 	correlation_df <- apply(X = dmr_gene_annotation_culled, MARGIN = 1, FUN = correlator)
	cl <- makeCluster(no_cores)
	if(!is.null(seed)){
		clusterSetRNGStream(cl, seed)
	}
	clusterExport(cl, varlist = c("dmr_data", "expr_data"), envir = env)
	clusterEvalQ(cl, expr = library(boot))
	correlation_mat <- t(parApply(cl = cl, X = dmr_gene_annotation_culled, MARGIN = 1, FUN = correlator))
	stopCluster(cl)
	correlation_mat <- data.frame(dmr_gene_annotation_culled, correlation_mat)
	# Found the following on https://stackoverflow.com/questions/31021868/merge-the-rows-in-r-with-the-same-row-name-concatenating-the-content-in-the-colu. The only issue is that it separates entries with a comma, which is unhandy if I want to save the table as a .csv.
	correlation_mat <- aggregate(contrast~., correlation_mat, FUN = toString)
	correlation_mat$contrast <- gsub(", *", ";", correlation_mat$contrast)
	#correlation_mat <- correlation_mat[order(abs(correlation_mat$rho), decreasing = T),]
	#correlation_mat <- correlation_mat[order(correlation_mat$rhoCI95_upper-correlation_mat$rhoCI95_lower), ]
	correlation_mat <- correlation_mat[order(correlation_mat$pval),]

	return(correlation_mat)
}

dmen_corr <- eqtm_func(dmen_betas_mean_df, expr_data, dmen_beta_gene, seed = 1)
dmen_corr$feature <- "enhancer"
dmen_corr_df <- data.frame(dmrChr = gsub("^(.+):[0-9]+-[0-9]+$", "\\1", dmen_corr$dmrcoords),
			   dmrStart = gsub("^.+:([0-9]+)-[0-9]+$", "\\1", dmen_corr$dmrcoords),
			   dmrEnd = gsub("^.+:[0-9]+-([0-9]+)$", "\\1", dmen_corr$dmrcoords),
			   enhChr = gsub("^(.+):[0-9]+-[0-9]+$", "\\1", dmen_corr$enhcoords),
			   enhStart = gsub("^.+:([0-9]+)-[0-9]+$", "\\1", dmen_corr$enhcoords),
			   enhEnd = gsub("^.+:[0-9]+-([0-9]+)$", "\\1", dmen_corr$enhcoords),
			   dmen_corr[,c(3:ncol(dmen_corr))])

dmprom_corr <- eqtm_func(dmprom_betas_mean_df, expr_data, dmprom_beta_gene, seed = 1)
dmprom_corr$feature <- "promoter"
dmprom_corr_df <- data.frame(dmrChr = gsub("^(.+):[0-9]+-[0-9]+$", "\\1", dmprom_corr$dmrcoords),
			   dmrStart = gsub("^.+:([0-9]+)-[0-9]+$", "\\1", dmprom_corr$dmrcoords),
			   dmrEnd = gsub("^.+:[0-9]+-([0-9]+)$", "\\1", dmprom_corr$dmrcoords),
			   promChr = gsub("^(.+):[0-9]+-[0-9]+$", "\\1", dmprom_corr$promcoords),
			   promStart = gsub("^.+:([0-9]+)-[0-9]+$", "\\1", dmprom_corr$promcoords),
			   promEnd = gsub("^.+:[0-9]+-([0-9]+)$", "\\1", dmprom_corr$promcoords),
			   dmprom_corr[,c(3:ncol(dmprom_corr))])
	

write.csv(dmen_corr_df, "dmenhancers_eqtm.csv")
write.csv(dmprom_corr_df, "dmpromoters_eqtm.csv")
```

## Visualization of the eQTM
```{r eQTM visualization}
eqtm_plot <- function(dmr_data, expr_data, coordinates, ensembl_id, degrees, plot_title, legend_title = NULL){
	#dmr_data: A dataframe containing the methylation data with individual methylation features as rows and samples as columns.
	#expr_data: A dataframe containing the expression data with individual expression features as rows and samples as columns.
	#degrees: A vector containing the degrees used to color the points with. Names must be the same as the ones used for the methylation and expression data.

	overlapping_samples <- intersect(colnames(expr_data), colnames(dmr_data)) 

	degrees <- degrees[overlapping_samples]
	dmr_data <- unlist(dmr_data[coordinates, overlapping_samples])
	expr_data <- unlist(expr_data[ensembl_id, overlapping_samples])

	plot_df <- data.frame(methylation = dmr_data, transcription = expr_data, groups = degrees) 

	require(ggplot2)
	p <- ggplot(plot_df, aes(x = methylation, y = transcription)) + 
		geom_point(aes(fill = groups), color = "black", shape = 21, size = 5) + 
		theme_bw() +
		xlab("Methylation") + 
		ylab("Expression") + 
		xlim(0,1) +
		ggtitle(plot_title) + 
		theme(plot.title = element_text(face = "bold"),
			axis.title = element_text(size = 14, face = "bold"),
			axis.text = element_text(size = 12), 
			legend.title = element_text(size = 14, face = "bold"),
			legend.text = element_text(size = 12),
			legend.position = "bottom")
	if(is.numeric(plot_df$groups) & !is.null(legend_title)){
		p <- p + scale_fill_continuous(name = legend_title)
	} else{
		p <- p + scale_fill_discrete(name = legend_title)
	}
	return(p)
}

named_swelling <- pData(fit2)$Swelling
names(named_swelling) <- pData(fit2)$Sample
named_sjc <- pData(fit2)$SJC
names(named_sjc) <- pData(fit2)$Sample

eqtm_top_plot <- function(dmr_data, expr_data, eqtm_top, degrees, number = 1, legend_title){
	if(length(number) == 1){
		name_plot <- paste0(eqtm_top[number, "dmrcoords"], "_", eqtm_top[number, "ensembl"], " (", eqtm_top[number, "symbol"], ")")
		png(file = paste0(eqtm_top[number, "ensembl"], ".png"), width = 800, height = 800, res = 150)
		print(eqtm_plot(dmr_data = dmr_data, expr_data = expr_data, coordinates = as.character(eqtm_top[number, "dmrcoords"]), degrees = degrees, ensembl_id = as.character(eqtm_top[number, "ensembl"]), plot_title = name_plot, legend_title = legend_title))
		dev.off()
	} else{
		lapply(number, function(iterator){ 
			name_plot <- paste0(eqtm_top[iterator, "dmrcoords"], "_", eqtm_top[iterator, "ensembl"], " (", eqtm_top[iterator, "symbol"], ")")
			png(file = paste0(iterator, "_", eqtm_top[iterator, "ensembl"], ".png"), width = 800, height = 800, res = 150)
			print(eqtm_plot(dmr_data = dmr_data, expr_data = expr_data, coordinates = as.character(eqtm_top[iterator, "dmrcoords"]), degrees = degrees, ensembl_id = as.character(eqtm_top[iterator, "ensembl"]), plot_title = name_plot, legend_title = legend_title))
			dev.off()
		})
	}
}

# Top 10 eQTMs
eqtm_top_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, eqtm_top = dmen_corr, degrees = named_sjc, number = 1:10, legend_title = "SJC")
eqtm_top_plot(dmr_data = dmprom_betas_mean_df, expr_data = expr_data, eqtm_top = dmprom_corr, degrees = named_sjc, number = 1:10, legend_title = "SJC")

# Understand the distribution of (p-values and) Spearman correlation coefficients.
rho_sig_plot <- function(corr_df, celltypes){
	require(ggExtra)
	require(ggplot2)
	rho_sig_plotter <- function(cell, corr_df){
			cell_df <- corr_df[grep(paste0(cell), corr_df$celltype), c("pval", "rho")]
#			plot_obj <- ggplot(cell_df, aes(x = rho, y = pval)) +
#				geom_point() + 
#				theme_bw() +
#				ggtitle(cell) + 
#				xlim(-1.2, 1.2) + 
#				ylim(-0.2, 1.2) + 
#				geom_rug()
#
#			dir.create("PvRho")
#			png(file = paste0("PvRho/", cell, ".png"), width = 800, height = 800, res = 150)
#			print(plot_obj)
#			dev.off()

			dir.create("Rho")
			hist_obj <- ggplot(cell_df, aes(x = rho)) +
				xlim(-1, 1) + 
				ylim(0, 2.5) +
				geom_histogram(aes(y=..density..),
						binwidth = 0.05,
						colour = "black", 
						fill = "white") +
				geom_density(alpha = 0.2) +
				ggtitle(cell) +
				theme_bw() 
			png(file = paste0("Rho/", cell, ".png"), width = 800, height = 800, res = 150)
			print(hist_obj)
			dev.off()
	}
	lapply(celltypes, function(cell){rho_sig_plotter(cell = cell, corr_df = corr_df)})
}

enhancer_celltypes <- unique(unlist(strsplit(as.character(dmen_corr$celltype), ";")))
rho_sig_plot(corr_df = dmen_corr, celltypes = enhancer_celltypes)
promoter_celltypes <- unique(gsub("_(C|V)B", "", unlist(strsplit(as.character(dmprom_corr$celltype), ";"))))
rho_sig_plot(corr_df = dmprom_corr, celltypes = promoter_celltypes)

plot_df <- dmen_corr %>% separate_rows(celltype, sep = "(,|;)")
png(file = "enhancer_rho_dist.png", width = 800, height = 800, res = 150)
ggplot(plot_df, aes(x = rho)) +
	geom_density(aes(fill = factor(celltype), alpha = 0.05)) +
	xlim(-1,1) +
	theme(legend.position = "bottom") +
	ggtitle("Enhancer rho distribution")
dev.off()

promoter_celltypes <- unique(unlist(strsplit(as.character(dmprom_corr$celltype), ";")))
rho_sig_plot(corr_df = dmprom_corr, celltypes = promoter_celltypes)

dmprom_corr2 <- dmprom_corr
dmen_corr2 <- dmen_corr
colnames(dmprom_corr2)[2] <- colnames(dmen_corr2)[2] <- "regcoords" 
dmreg_corr <- rbind(dmprom_corr2, dmen_corr2)
```

# Specific questions to be answered
Now that we have the eQTM analysis results, we have a few questions of interest regarding the promoters and enhancers:
1. Which genes are regulated by several regulatory regions?
2. Which regulatory regions contain DMRs and affect multiple genes?

## Question 1: Which genes are regulated by several regulatory regions?
To answer this question, the fast way is to simply tabulate the number of gene symbol occurrences in the list
```{r Gene symbol summary}
table(dmreg_corr$symbol)
table(table(dmreg_corr$symbol))
```
At first glance it appears as though the differentially methylated regulatory regions appear to have many genes that are regulated by several regulatory elements, which is what we expect. However, the picture is a bit exaggerated as we see that some DMRs are "non-unique". This was not picked up because the DMRs map to the same region but are not exactly the same. This also means that the associated statistics are slightly different. 
Our solution is to remove the duplicated entries per gene symbol/ensembl ID. While removing multiple entries for the same target gene symbol is not an ideal situation (i.e. we accidentally remove multiple non-overlapping DMRs within the same region), it gives us at the very least an idea of which genes are affected by DMRs in different regulatory regions. The entry that is removed right now is the entry that comes first. Later updates will have to change this into something more agreeable (i.e. longest/shortest DMR).
```{r Deduping overlapping regions}
dmreg_corr_promgene <- split(dmreg_corr, dmreg_corr$symbol)
dmreg_corr_promgene <- dmreg_corr_promgene[lapply(dmreg_corr_promgene, nrow) > 1]
dmreg_corr_promgene_df <- do.call(rbind, dmreg_corr_promgene)

dmreg_corr_promgene_filtered <- lapply(dmreg_corr_promgene, function(promgene){
	       reg_ens <- paste0(promgene$regcoords, promgene$ensembl)
	       deduped <- promgene[!duplicated(reg_ens), ] # This is NOT an ideal solution
	       deregduped <- aggregate(regcoords~., deduped, FUN = toString) 
	       return(deregduped)
})
dmreg_corr_promgene_filtered <- dmreg_corr_promgene_filtered[lapply(dmreg_corr_promgene_filtered, nrow) > 1]

dmreg_corr_promgene_filtered_df <- do.call(rbind, dmreg_corr_promgene_filtered)
dmreg_corr_promgene_filtered_df <- aggregate(regcoords~., dmreg_corr_promgene_filtered_df, FUN = toString)

write.csv(dmreg_corr_promgene_df, "promgene_raw.csv")
write.csv(dmreg_corr_promgene_filtered_df, "promgene_filtered.csv")

promgenes <- names(which(sort(table(unique(dmreg_corr[,c("regcoords", "ensembl", "symbol")])[,"symbol"]))>1))
dmreg_corr_promgene_uniqueregs_df <- dmreg_corr_promgene_df[dmreg_corr_promgene_df$symbol %in% promgenes,]
write.csv(dmreg_corr_promgene_uniqueregs_df, "promgene_uniqueregs.csv")

```
A DMR was found for BRD7 is doing
```{r BRD7}
png("BRD7_enhancer.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "16:50203607-50205182", "ENSG00000166164", degrees = named_sjc, plot_title = "BRD7_enhancer", legend_title = "SJC"))
dev.off()
```
Two DMRs were found in one enhancer of KAT6B, one of which was strongly correlated
```{r KAT6B}
png("KAT6B_enhancer.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "10:75928975-75929471", "ENSG00000156650", degrees = named_sjc, plot_title = "KAT6B_enhancer", legend_title = "SJC"))
dev.off()
```


Analysis showed a very interesting correlation for PAPD5, which was found to be regulated through its promoter and an enhancer. 
```{r PAPD5}
#Enhancer
png("PAPD5_enhancer.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "16:50203607-50205157", "ENSG00000121274", degrees = named_sjc, plot_title = "PAPD5_enhancer", legend_title = "SJC"))
dev.off()
#Promoter
png("PAPD5_promoter.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmprom_betas_mean_df, expr_data = expr_data, coordinates = "16:50154665-50154991", "ENSG00000121274", degrees = named_sjc, plot_title = "PAPD5_promoter", legend_title = "SJC"))
dev.off()
```

Another interesting hit was found for SP140, where we found two distinct DMRs within the same enhancer
```{r SP140}
png("SP140_1.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "2:231541887-231541958", "ENSG00000079263", degrees = named_sjc, plot_title = "SP140_enhancer1", legend_title = "SJC"))
dev.off()
#Promoter
png("SP140_2.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmprom_betas_mean_df, expr_data = expr_data, coordinates = "2:231545170-231546134", "ENSG00000079263", degrees = named_sjc, plot_title = "SP140_enhancer2", legend_title = "SJC"))
dev.off()

```


Despite our best efforts we managed to find a DMR that resided within an enhancer for two genes, thereby answering question 2 partially. Surprisingly both genes appear to behave the same.
```{r DMR 1:181131123-181132025}
png("MR1.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "1:181131123-181132025", "ENSG00000153029", factors = named_sjc, plot_name = "MR1", legend_name = "SJC"))
dev.off()
png("IER5.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "1:181131123-181132025", "ENSG00000162783", factors = named_sjc, plot_name = "IERS", legend_name = "SJC"))
dev.off()

#Need to correlate the actual expression of MR1 and IER5
IER5_MR1 <- t(expr_data[c("ENSG00000153029", "ENSG00000162783"), ] )
summary(lm(IER5_MR1[,1] ~ IER5_MR1[,2]))
```

## Question 2: Which regulatory regions contain DMRs and affect multiple genes?
To answer the question which differentially methylated regulatory regions affect multiple genes, we can utilize the same strategy as the one implemented in question 1, namely sorting and splitting for the regulatory coordinates. 
```{r Regulatory region summary}
table(dmreg_corr$regcoords)
table(table(dmreg_corr$regcoords))
```

```{r Finding the promiscuous DMregs}
dmreg_corr_promreg <- split(dmreg_corr, dmreg_corr$regcoords)
dmreg_corr_promreg <- dmreg_corr_promreg[which(lapply(dmreg_corr_promreg, nrow) > 1)]
dmreg_corr_promreg <- dmreg_corr_promreg[order(unlist(lapply(dmreg_corr_promreg, nrow)))]

dmreg_corr_promreg_df <- do.call(rbind, dmreg_corr_promreg)

write.csv(dmreg_corr_promreg_df, "prom_dmreg_raw.csv") 
```

One of the interesting DMregs was found at 14:23366500-23366804. It is associated to SLC7A7, which was reported in association to RA previously.
```{r SLC7A7}
png("SLC7A7.png", width = 800, height = 800, res = 150)
print(eqtm_plot(dmr_data = dmen_betas_mean_df, expr_data = expr_data, coordinates = "14:23366500-23366804", "ENSG00000155465", factors = named_sjc, plot_name = "SLC7A7", legend_name = "SJC"))
dev.off()
```
